gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java

```diff
@@ -0,0 +1,322 @@
+          .append();
+    } catch (NoSuchTableException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static Iterator<Row> getContentRows(
+      Iterator<IdentifiedResult> result, String runID, Timestamp startedAt) {
+    List<Row> rows = new ArrayList<>();
+    ObjectMapper objectMapper = new ObjectMapper();
+    while (result.hasNext()) {
+      result
+          .next()
+          .getContentValues()
+          .forEach(
+              (refName, value) ->
+                  value.forEach(
+                      (contentId, contentValues) ->
+                          rows.add(
+                              getContentRow(
+                                  runID,
+                                  startedAt,
+                                  refName,
+                                  contentId,
+                                  contentValues,
+                                  objectMapper))));
+    }
+    return rows.iterator();
+  }
+
+  private static Row getMarkerRow(String runID, Timestamp startedAt) {
+    return RowFactory.create(
+        IdentifiedResultsRepo.TYPE_GC_MARKER, startedAt, runID, null, null, null);
+  }
+
+  private static Row getContentRow(
+      String runID,
+      Timestamp startedAt,
+      String refName,
+      String contentId,
+      ContentValues contentValues,
+      ObjectMapper objectMapper) {
+    return RowFactory.create(
+        IdentifiedResultsRepo.TYPE_CONTENT,
+        startedAt,
+        runID,
+        contentId,
+        contentValues.getExpiredContents().stream()
+            .map(
+                content -> {
+                  try {
+                    // serialize to json and store the contents
+                    return objectMapper.writeValueAsString(content);
+                  } catch (JsonProcessingException e) {
+                    throw new RuntimeException(e);
+                  }
+                })
+            .toArray(String[]::new),
+        refName);
+  }
+
+  private static StructType checkAndCreateTable(
+      SparkSession sparkSession,
+      String catalogName,
+      TableIdentifier tableIdentifier,
+      String catalogAndTable,
+      String gcRefName) {
+    // get Nessie catalog
+    Catalog nessieCatalog =
+        CatalogUtil.loadCatalog(
+            NessieCatalog.class.getName(),
+            catalogName,
+            catalogConfWithRef(sparkSession, catalogName, gcRefName),
+            sparkSession.sparkContext().hadoopConfiguration());
+    // create table if not exists
+    sparkSession.sql(String.format(SCHEMA_DDL, withRefName(catalogAndTable, gcRefName)));
+    // get schema
+    return SparkSchemaUtil.convert(nessieCatalog.loadTable(tableIdentifier).schema());
+  }
+
+  private static void addRowToIdentifiedResults(
+      IdentifiedResult identifiedResult, ObjectMapper objectMapper, Row row) {
+    List<Object> contentsInRow = row.getList(1);
+    List<Content> contents =
```

Comment:
why materialize teh List just to do a foreach on it?

Reply:
yeah, no need.